### PR TITLE
A Layout of the JupyterHub repository

### PR DESCRIPTION
**Pull request by Temidayo32**
@choldgraf raised an issue #3080 on the need for a documentation that provides a layout of the Jupyterhub repository. This pull request addresses the issue. Please, let me know of any further corrections to the made.

I am also an Outreachy intern. This effectively serves as my first contribution to the project during this contribution period.

### CODE REVIEW

File: docs/source/contributing/layout.rst [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+share/jupyterHub/
+    This directory contains JupyterHub data and static files.
+
+singleuser/
+    This directory contains the Dockerfile for building a JupyterHub image for a single user. You can check out the `README` in the directory for more details.
+
+testing/
+    This directory contains a sample JupyterHub configuration file for testing.
+
+Executable files
```
Comment (minrk) [inline docs/source/contributing/layout.rst:44]: These aren't executables, but I'm not sure what exactly the best name is. Configuration files? Packaging files?

File: docs/source/contributing/layout.rst [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    This directory contains the Dockerfile for building the JupyterHub base image. You can check out the `README` in the directory for more details.
+
+docs/
+    This directory contains the documentation files for the JupyterHub project.
+
+examples/
+    This directory details several services provided by the JupyterHub project, alongside their configurations to enable you configure JupyterHub to use those services.
+
+jsx/
+    This directory contains current updates to the JupyterHub Admin Dashboard.
```
Comment (minrk) [inline docs/source/contributing/layout.rst:27]: These are the "javascript sources for the admin dashboard"

File: docs/source/contributing/layout.rst [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    A requirement for using setuptools(a build backend) in building the JupyterHub distribution package.
+
+MANIFEST.in
+    A file for instructing setuptools on the files to add or remove when building the JupyterHub distribution package (sdist).
+
+requirements.txt
+    This file lists the dependencies of the JupyterHub codebase.
+
+dev-requirements.txt
+    This file lists additional dependencies of the JupyterHub codebase.
```
Comment (minrk) [inline docs/source/contributing/layout.rst:72]: These are the dependencies used to develop/test jupyterhub, as opposed to "additional" dependencies.

File: docs/source/contributing/layout.rst [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    Another components file used in packaging JupyterHub. It stages the frontend dependencies of the JupyterHub repository from the node_modules directory into the components sub-directory. Check the static directory in the `share/jupyterHub` directory to find the components sub-directory.
+
+package.json
+    This file contains information on JupyterHub nodejs dependencies.
+
+pytest.ini
+    A configuration file for customizing Pytest behaviors to suit the usage of Pytest in the JupyterHub project. Pytest is a python tool for running tests.
+
+pyproject.toml
+    A requirement for using setuptools(a build backend) in building the JupyterHub distribution package.
```
Comment (minrk) [inline docs/source/contributing/layout.rst:63]: pyproject.toml is actually used to configure a few tools. Specifically, code formatting and release tools (tbump).

File: docs/source/contributing/layout.rst [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    This file contains the JupyterHub license information. It explains the permission you have using source code from the repository.
+
+RELEASE.md
+    This file provides information on making releases of the JupyterHub project.
+
+SECURITY.md
+    This file provides information on how to report an identified security vulnerability in the JupyterHub project.
+
+
+dotfiles
```
Comment (minrk) [inline docs/source/contributing/layout.rst:99]: dotfiles aren't a separate category - these are mostly the same as what's called 'executable files' right now. Some of them just happen to start with '.'

**Review by minrk — COMMENTED** [review]
Thanks @Temidayo32, I think this is useful!

I think when I mentioned a layout of the repo in https://github.com/jupyterhub/team-compass/issues/296#issuecomment-635826046 I was referring to how the _code_ (within `jupyterhub/` is organized (e.g. `orm.py` contains the database code, the REST API is in `jupyterhub/apihandlers`, etc.), but this is still useful.

@choldgraf @willingc what do you think is the right level of detail to describe "all the files" of a single repo? I don't see this very often in repo docs.

**Comment by manics** [comment]
I think the main thing to consider is whether this will be kept up to date. @willingc [suggested a top-level `LAYOUT` file](https://github.com/jupyterhub/jupyterhub/issues/3080#issuecomment-643657778) which has the advantage it's more visible to anyone modifying the code.

If we keep it as RST then how about hyperlinking the files docs to the `main` branch, so if anything moves in future we can use a linkchecker to detect a broken link?

File: docs/source/contributing/layout.rst (hunk 3 above)
Comment (Temidayo32) [reply docs/source/contributing/layout.rst:72]: Got it

File: docs/source/contributing/layout.rst (hunk 1 above)
Comment (Temidayo32) [reply docs/source/contributing/layout.rst:44]: @minrk  the naming of the heading is actually based off a similar documentation written  by @willingc 
https://github.com/openhatch/oh-mainline/blob/master/LAYOUT#L4

Comment (Temidayo32) [reply docs/source/contributing/layout.rst:44]: And yeah, I could still do that if you want. A layout of the code in `jupyterhub/`. I would just have it added as  subs under the `jupyterhub` dir currently in the layout. What do you think?

**Comment by Temidayo32** [comment]
@manics Yeah, that is a good observation. I would think if kept high level enough, then updating needs would be minimal. The documentation would go through a major revamp on the completion of the Outreachy intern. This could mean a major modification of the files in the docs directory. However, if the layout is kept just describing the docs folder and not going indepth to describe its sub-folder, then the modification that would take place would not require an updating of the Layout

**Comment by Temidayo32** [comment]
This pull request contains commits that fixed the issue #3875

File: docs/source/contributing/layout.rst (hunk 1 above)
Comment (Temidayo32) [reply docs/source/contributing/layout.rst]: Hi @minrk I just made a new commit as regards this https://github.com/jupyterhub/jupyterhub/pull/4066/commits/4eb13d5eafac1af925ce310ff7d244d239c42dbf

**Comment by yuvipanda** [comment]
Given it's been a couple years, I'm going to close this one. Please re-open if you disagree.

**Pull request closed without merging**